libsel4vm/src/arch/arm/vgic/vgic_v2.c

```diff
@@ -0,0 +1,300 @@
+    struct virq_handle *virq = vgic_irq_dequeue(vgic, vcpu);
+    if (virq) {
+        return vgic_vcpu_load_list_reg(vgic, vcpu, idx, 0, virq);
+    }
+
+    return 0;
+}
+
+
+static void vgic_dist_reset(struct vgic_dist_device *d)
+{
+    struct gic_dist_map *gic_dist;
+    gic_dist = vgic_priv_get_dist(d);
+    memset(gic_dist, 0, sizeof(*gic_dist));
+    gic_dist->ic_type         = 0x0000fce7; /* RO */
+    gic_dist->dist_ident      = 0x0200043b; /* RO */
+
+    for (int i = 0; i < CONFIG_MAX_NUM_NODES; i++) {
+        gic_dist->enable_set0[i]   = 0x0000ffff; /* 16bit RO */
+        gic_dist->enable_clr0[i]   = 0x0000ffff; /* 16bit RO */
+    }
+
+    /* Reset value depends on GIC configuration */
+    gic_dist->config[0]       = 0xaaaaaaaa; /* RO */
+    gic_dist->config[1]       = 0x55540000;
+    gic_dist->config[2]       = 0x55555555;
+    gic_dist->config[3]       = 0x55555555;
+    gic_dist->config[4]       = 0x55555555;
+    gic_dist->config[5]       = 0x55555555;
+    gic_dist->config[6]       = 0x55555555;
+    gic_dist->config[7]       = 0x55555555;
+    gic_dist->config[8]       = 0x55555555;
+    gic_dist->config[9]       = 0x55555555;
+    gic_dist->config[10]      = 0x55555555;
+    gic_dist->config[11]      = 0x55555555;
+    gic_dist->config[12]      = 0x55555555;
+    gic_dist->config[13]      = 0x55555555;
+    gic_dist->config[14]      = 0x55555555;
+    gic_dist->config[15]      = 0x55555555;
+
+    /* Configure per-processor SGI/PPI target registers */
+    for (int i = 0; i < CONFIG_MAX_NUM_NODES; i++) {
+        for (int j = 0; j < ARRAY_SIZE(gic_dist->targets0[i]); j++) {
+            for (int irq = 0; irq < sizeof(uint32_t); irq++) {
+                gic_dist->targets0[i][j] |= ((1 << i) << (irq * 8));
+            }
+        }
+    }
+    /* Deliver the SPI interrupts to the first CPU interface */
+    for (int i = 0; i < ARRAY_SIZE(gic_dist->targets); i++) {
+        gic_dist->targets[i] = 0x1010101;
+    }
+
+    /* identification */
+    gic_dist->periph_id[4]    = 0x00000004; /* RO */
+    gic_dist->periph_id[8]    = 0x00000090; /* RO */
+    gic_dist->periph_id[9]    = 0x000000b4; /* RO */
+    gic_dist->periph_id[10]   = 0x0000002b; /* RO */
+    gic_dist->component_id[0] = 0x0000000d; /* RO */
+    gic_dist->component_id[1] = 0x000000f0; /* RO */
+    gic_dist->component_id[2] = 0x00000005; /* RO */
+    gic_dist->component_id[3] = 0x000000b1; /* RO */
+}
+
+int vm_register_irq(vm_vcpu_t *vcpu, int irq, irq_ack_fn_t ack_fn, void *cookie)
+{
+    struct vgic *vgic = vgic_device_get_vgic(vgic_dist);
+    assert(vgic);
+
+    struct virq_handle *virq_data = calloc(1, sizeof(*virq_data));
+    if (!virq_data) {
+        return -1;
+    }
+
+    virq_init(virq_data, irq, ack_fn, cookie);
+
+    int err = virq_add(vcpu, vgic, virq_data);
+    if (err) {
+        free(virq_data);
+        return -1;
+    }
+
+    return 0;
+}
+
+int vm_inject_irq(vm_vcpu_t *vcpu, int irq)
```

Comment:
It seem this function remains the same in GICv2 and GICv3. Thus I think it would make sense to keep thios in a common file and avoid code duplication. If there are differences in the future, we can still split this up,